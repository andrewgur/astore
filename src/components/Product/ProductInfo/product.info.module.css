.productInfo__title {
  font-size: 20px;
  line-height: 1.35;
  font-weight: 400;
  margin: 0 0 20px;
}
.productInfo__price {
  font-size: 20px;
  line-height: 1.35;
  font-weight: 700;
}
.productInfo__label {
  display: block;
  padding-bottom: 3px;
  font-size: 12px;
  line-height: 1.55;
}
.productInfo__selectBlock {
  margin: 10px 0;
}
.productInfo__field {
  margin: 0 0 15px;
}
.productInfo__select {
  font-size: 12px;
  line-height: 1.55;
  width: 100%;
  border: 1px #ddd solid;
  background: #f8f8f8;
  color: #000;
  box-sizing: border-box;
  cursor: pointer;
  padding: 5px 30px 5px 10px;
  border-radius: 5px;
  width: 150px;
}
.productInfo__button {
  font-family: 'StyreneA',Arial,sans-serif;
  margin: 20px 0 30px;
  border-radius: 30px;
  font-size: 14px;
  font-weight: 700;
}
.productInfo__description {
  font-size: 12px;
  line-height: 1.5;
  font-weight: 400;
}
.productInfo__form {
  margin: 40px 0 10px 0;
  max-width: 250px;
}

@media (max-width: 960px){
  .productInfo {
      margin: 40px auto 0;
      max-width: 560px;
  }
}