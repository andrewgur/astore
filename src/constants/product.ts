export const TITLE_SIZES = 'размер';
export const TITLE_COLORS = 'цвет';
export const TITLE_MODELS = 'модель';
export const TITLE_STICKER_NUMBERS = 'номер стикера';

export const TITLE_DESC = {
  sizes: TITLE_SIZES,
  colors: TITLE_COLORS,
  models: TITLE_MODELS,
  stickerNumbers: TITLE_STICKER_NUMBERS,
};