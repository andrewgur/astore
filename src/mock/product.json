{
  "id": 5,
  "preview": "http://qa-games.ru/astore/public/images/43306375.jpeg",
  "images": [
    "http://qa-games.ru/astore/public/images/43306375.jpeg",
    "http://qa-games.ru/astore/public/images/25133982.png",
    "http://qa-games.ru/astore/public/images/93661622.png",
    "http://qa-games.ru/astore/public/images/1_3d.png",
    "http://qa-games.ru/astore/public/images/2_3d.png",
    "http://qa-games.ru/astore/public/images/45157942.png",
    "http://qa-games.ru/astore/public/images/Frame_118.png"
  ],
  "title": "Худи с бархатными стикерами",
  "subtitle": "Выберите один из восьми стикеров",
  "price": 4199,
  "description": "Выберите стикер, а мы перенесём его на ткань — как на фото. Одежду можно стирать в машинке при 30 °C, вывернув наизнанку. Гладить — с внутренней стороны. Посмотреть и потрогать все стикеры можно в A-Store на Технопарке. А ещё там можно добавить сразу несколько стикеров на одну вещь.",
  "colors": [
    "white",
    "black",
    "red"
  ],
  "sizes": [
    "XS",
    "S",
    "M",
    "L",
    "XL"
  ],
  "stickerNumbers": [
    1,
    2,
    3,
    4,
    5,
    6,
    7,
    8
  ],
  "availability": true
}