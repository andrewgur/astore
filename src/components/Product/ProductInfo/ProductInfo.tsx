import { FC } from 'react';
import { Typography } from '@alfalab/core-components/typography';
import { Button } from '@alfalab/core-components/button';
import { FormikValues, useFormik } from 'formik';
import { SelectResponsive } from '@alfalab/core-components/select/responsive';
import { OptionShape } from '@alfalab/core-components/select/typings';
import { ProductCartOptions } from 'types/Product';
import cls from './product.info.module.css';
import { TITLE_COLORS, TITLE_MODELS, TITLE_SIZES, TITLE_STICKER_NUMBERS } from 'constants/product';

type ProductInfoProps = {
  title: string;
  price: number;
  description: string;
  stickerNumbers?: OptionShape[];
  models?: OptionShape[];
  colors?: OptionShape[];
  sizes?: OptionShape[];
  handleAdd: (data: ProductCartOptions) => void;
};

type SelectListItemType = {
  name: 'colors' | 'stickerNumbers' | 'models' | 'colors' | 'sizes';
  label: string;
  options: OptionShape[] | undefined;
};

export type FormDataType = {
  stickerNumbers?: OptionShape;
  models?: OptionShape;
  colors?: OptionShape;
  sizes?: OptionShape;
};

export const ProductInfo: FC<ProductInfoProps> = ({ title, price, description, stickerNumbers, models, colors, sizes, handleAdd }) => {

  const selectList: SelectListItemType[] = [
    {
      name: 'colors',
      label: TITLE_COLORS,
      options: colors
    },
    {
      name: 'stickerNumbers',
      label: TITLE_STICKER_NUMBERS,
      options: stickerNumbers
    },
    {
      name: 'models',
      label: TITLE_MODELS,
      options: models
    },
    {
      name: 'sizes',
      label: TITLE_SIZES,
      options: sizes
    },
  ];

  const initialValues = selectList.reduce((acc: FormDataType, item: SelectListItemType) => {
    if (item.options?.length) {
      acc[item.name] = item.options[0];
    }
    return acc;
  }, {});

  const formik = useFormik({
    initialValues,
    onSubmit: (values: FormikValues) => {
      const data = Object.entries(values)
        .reduce((acc: ProductCartOptions, item) => {
          acc[item[0] as keyof ProductCartOptions] = item[1].content;
          return acc;
        }, {});
      handleAdd(data);
    }
  });

  return (
    <div className={cls.productInfo}>
      <Typography.Title className={cls.productInfo__title} tag='h1'>{title}</Typography.Title>
      <div className={cls.productInfo__price}>{price} ₽</div>
      <form className={cls.productInfo__form} onSubmit={formik.handleSubmit}>
        {selectList.map(item => {
          if (item.options?.length) {
            return (
              <div key={item.name} className={cls.productInfo__field}>
                <SelectResponsive
                  dataTestId={item.name}
                  key={item.name}
                  options={item.options}
                  block={true}
                  name={item.name}
                  label={item.label}
                  labelView='outer'
                  optionsSize='s'
                  onChange={(option) => formik.setFieldValue(item.name, option.selected)}
                  selected={formik.values[item.name]}
                />
              </div>
            );
          }
        }
        )}
        <Button
          className={cls.productInfo__button}
          type='submit'
          view='primary'
          dataTestId='product-add'
        >
          В корзину
        </Button>
      </form>
      <Typography.Text className={cls.productInfo__description} tag='p'>{description}</Typography.Text>
    </div>
  );
};