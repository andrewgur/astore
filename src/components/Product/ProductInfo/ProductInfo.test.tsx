import React from 'react';
import { render, screen } from '@testing-library/react';
import { ProductInfo } from './ProductInfo';
import product from './../../../mock/product.json';

describe('ProductGallery Component', () => {

  const { title, price, description, stickerNumbers, colors, sizes } = product;

  const setup = () => render(
    <ProductInfo
      title={title}
      price={price}
      description={description}
      colors={colors}
      sizes={sizes}
      models={[]}
      stickerNumbers={stickerNumbers}
    />
  );

  it('should render props', () => {
    setup();
    const titleElem = screen.getByText(title);
    expect(titleElem.textContent).toBe(title);

    const descrElem = screen.getByText(description);
    expect(descrElem.textContent).toBe(description);

    const priceElem = screen.getByText(price, { exact: false });
    expect(priceElem).toBeInTheDocument();

    ['colors', 'sizes', 'stickerNumbers'].map(name => {
      const element = screen.getByRole('combobox', { name: name });
      expect(element).toBeInTheDocument();
    });

  });

  it('shouldn\'t redner props', () => {
    setup();
    const element = screen.queryByRole('combobox', { name: 'models' });
    expect(element).not.toBeInTheDocument();
  });
});