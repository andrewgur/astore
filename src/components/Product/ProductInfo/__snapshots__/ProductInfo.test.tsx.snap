// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductGallery Component should redner props 1`] = `
<div>
  <div
    class="productInfo"
  >
    <h1
      class="typography__component_1c381 productInfo__title typography__styrene-medium_17h0l typography__medium_1c381"
    />
    <div
      class="productInfo__price"
    >
       ₽
    </div>
    <form
      action="#"
    >
      <button
        class="button__component_4vwd7 button__m_4vwd7 button__component_re924 button__primary_re924 productInfo__button"
        type="submit"
      >
        <span
          class="button__text_4vwd7 button__stretchText_4vwd7"
        >
          В корзину
        </span>
      </button>
    </form>
    <p
      class="typography__paragraphWithMargins_9xfoj productInfo__description typography__primary-medium_9xfoj"
    />
  </div>
</div>
`;

exports[`ProductGallery Component shouldn't redner props 1`] = `
<div>
  <div
    class="productInfo"
  >
    <h1
      class="typography__component_1c381 productInfo__title typography__styrene-medium_17h0l typography__medium_1c381"
    />
    <div
      class="productInfo__price"
    >
       ₽
    </div>
    <form
      action="#"
    >
      <button
        class="button__component_4vwd7 button__m_4vwd7 button__component_re924 button__primary_re924 productInfo__button"
        type="submit"
      >
        <span
          class="button__text_4vwd7 button__stretchText_4vwd7"
        >
          В корзину
        </span>
      </button>
    </form>
    <p
      class="typography__paragraphWithMargins_9xfoj productInfo__description typography__primary-medium_9xfoj"
    />
  </div>
</div>
`;
